When a trait isn't implemented, but another similar impl is found, point at it

estebank opened the pull request:
```
error[E0277]: the trait bound `u32: Trait` is not satisfied
  --> $DIR/trait_objects_fail.rs:26:9
   |
LL |     foo(&10_u32);
   |         ^^^^^^^ the trait `Trait` is not implemented for `u32`
   |
help: the trait `Trait<12>` is not implemented for `u32`
      but trait `Trait<2>` is implemented for it
  --> $DIR/trait_objects_fail.rs:7:1
   |
LL | impl Trait<2> for u32 {}
   | ^^^^^^^^^^^^^^^^^^^^^
   = note: required for the cast from `&u32` to `&dyn Trait`
```

Pointing at the `impl` definition that *could* apply given a different self type is *particularly* useful when it has a blanket self type, as it might not be obvious and is not trivially greppable:

```
error[E0277]: the trait bound `RawImpl<_>: Raw<_>` is not satisfied
  --> $DIR/issue-62742.rs:4:5
   |
LL |     WrongImpl::foo(0i32);
   |     ^^^^^^^^^ unsatisfied trait bound
   |
help: the trait `Raw<_>` is not implemented for `RawImpl<_>`
      but trait `Raw<[_]>` is implemented for it
  --> $DIR/issue-62742.rs:29:1
   |
LL | impl<T> Raw<[T]> for RawImpl<T> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: required by a bound in `SafeImpl`
  --> $DIR/issue-62742.rs:33:35
   |
LL | pub struct SafeImpl<T: ?Sized, A: Raw<T>>(PhantomData<(A, T)>);
   |                                   ^^^^^^ required by this bound in `SafeImpl`
```

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

estebank committed:
When a trait isn't implemented, but another similar impl is found, point at it:

```
error[E0277]: the trait bound `u32: Trait` is not satisfied
  --> $DIR/trait_objects_fail.rs:26:9
   |
LL |     foo(&10_u32);
   |         ^^^^^^^ the trait `Trait` is not implemented for `u32`
   |
help: the trait `Trait<12>` is not implemented for `u32`
      but trait `Trait<2>` is implemented for it
  --> $DIR/trait_objects_fail.rs:7:1
   |
LL | impl Trait<2> for u32 {}
   | ^^^^^^^^^^^^^^^^^^^^^
   = note: required for the cast from `&u32` to `&dyn Trait`
```

Pointing at the `impl` definition that *could* apply given a different self type is particularly useful when it has a blanket self type, as it might not be obvious and is not trivially greppable:

```
error[E0277]: the trait bound `RawImpl<_>: Raw<_>` is not satisfied
  --> $DIR/issue-62742.rs:4:5
   |
LL |     WrongImpl::foo(0i32);
   |     ^^^^^^^^^ unsatisfied trait bound
   |
help: the trait `Raw<_>` is not implemented for `RawImpl<_>`
      but trait `Raw<[_]>` is implemented for it
  --> $DIR/issue-62742.rs:29:1
   |
LL | impl<T> Raw<[T]> for RawImpl<T> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: required by a bound in `SafeImpl`
  --> $DIR/issue-62742.rs:33:35
   |
LL | pub struct SafeImpl<T: ?Sized, A: Raw<T>>(PhantomData<(A, T)>);
   |                                   ^^^^^^ required by this bound in `SafeImpl`
```

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17086054794/job/48450293452) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/48450293452) [(plain)](https://github.com/rust-lang/rust/commit/3967b4e311151f95819b7b4abb3ef4a5ebecb2e9/checks/48450293452/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] tests/ui/impl-trait/diagnostics/highlight-difference-between-expected-trait-and-found-trait.rs stdout ----
Saved the actual svg to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/impl-trait/diagnostics/highlight-difference-between-expected-trait-and-found-trait/highlight-difference-between-expected-trait-and-found-trait.svg`
diff of svg:

- <svg width="1188px" height="470px" xmlns="http://www.w3.org/2000/svg">
+ <svg width="995px" height="470px" xmlns="http://www.w3.org/2000/svg">
2   <style>
3     .fg { fill: #AAAAAA }
4     .bg { background: #000000 }

30 </tspan>
31     <tspan x="10px" y="82px"><tspan class="fg-ansi256-012 bold">LL</tspan><tspan> </tspan><tspan class="fg-ansi256-012 bold">|</tspan><tspan> fn foo() -&gt; impl Foo&lt;i32&gt; {</tspan>
32 </tspan>
-     <tspan x="10px" y="100px"><tspan>   </tspan><tspan class="fg-ansi256-012 bold">|</tspan><tspan>             </tspan><tspan class="fg-ansi256-009 bold">^^^^^^^^^^^^^</tspan><tspan> </tspan><tspan class="fg-ansi256-009 bold">unsatisfied trait bound</tspan>
+     <tspan x="10px" y="100px"><tspan>   </tspan><tspan class="fg-ansi256-012 bold">|</tspan><tspan>             </tspan><tspan class="fg-ansi256-009 bold">^^^^^^^^^^^^^</tspan><tspan> </tspan><tspan class="fg-ansi256-009 bold">the trait `Bar&lt;i32&gt;` is not implemented for `Struct`</tspan>
34 </tspan>
35     <tspan x="10px" y="118px"><tspan class="fg-ansi256-012 bold">LL</tspan><tspan> </tspan><tspan class="fg-ansi256-012 bold">|</tspan><tspan>     Struct</tspan>
36 </tspan>


The actual svg differed from the expected svg
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args impl-trait/diagnostics/highlight-difference-between-expected-trait-and-found-trait.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/impl-trait/diagnostics/highlight-difference-between-expected-trait-and-found-trait.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/impl-trait/diagnostics/highlight-difference-between-expected-trait-and-found-trait" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--error-format=human" "--color=always"
stdout: none
--- stderr -------------------------------
error[E0277]: the trait bound `Struct: Foo<i32>` is not satisfied
  --> /checkout/tests/ui/impl-trait/diagnostics/highlight-difference-between-expected-trait-and-found-trait.rs:15:13
   |
LL | fn foo() -> impl Foo<i32> {
   |             ^^^^^^^^^^^^^ the trait `Bar<i32>` is not implemented for `Struct`
LL |     Struct
   |     ------ return type was inferred to be `Struct` here
   |
help: the trait `Bar<i32>` is not implemented for `Struct`
      but trait `Bar<()>` is implemented for it
  --> /checkout/tests/ui/impl-trait/diagnostics/highlight-difference-between-expected-trait-and-found-trait.rs:13:1
   |
LL | impl<'a> Bar<()> for Struct {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for that trait implementation, expected `()`, found `i32`
note: required for `Struct` to implement `Foo<i32>`
  --> /checkout/tests/ui/impl-trait/diagnostics/highlight-difference-between-expected-trait-and-found-trait.rs:10:12
   |
LL | impl<T, K> Foo<K> for T where T: Bar<K>
   |            ^^^^^^     ^          ------ unsatisfied trait bound introduced here

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0277`.
------------------------------------------
```

</details>

estebank committed:
When more than a single `impl` and less than 4 could apply, point at them

```
error[E0277]: the trait bound `[[u16; 3]; 2]: Bar` is not satisfied
  --> $DIR/issue-67185-2.rs:21:6
   |
LL | impl Foo for FooImpl {}
   |      ^^^ the trait `Bar` is not implemented for `[[u16; 3]; 2]`
   |
help: the following other types implement trait `Bar`
  --> $DIR/issue-67185-2.rs:9:1
   |
LL | impl Bar for [u16; 4] {}
   | ^^^^^^^^^^^^^^^^^^^^^ `[u16; 4]`
LL | impl Bar for [[u16; 3]; 3] {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ `[[u16; 3]; 3]`
note: required by a bound in `Foo`
  --> $DIR/issue-67185-2.rs:14:30
   |
LL | trait Foo
   |       --- required by a bound in this trait
LL | where
LL |     [<u8 as Baz>::Quaks; 2]: Bar,
   |                              ^^^ required by this bound in `Foo`
```

nnethercote commented:
Looks fine, r=me once everything is green.

bors commented:
:pushpin: Commit 898ddc3e3858f2195bb3cc75ec1d01d2caf8412d has been approved by `nnethercote`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nnethercote 898ddc3e3858f2195bb3cc75ec1d01d2caf8412d -->
<!-- homu: {"type":"Approved","sha":"898ddc3e3858f2195bb3cc75ec1d01d2caf8412d","approver":"nnethercote","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
The job **`aarch64-msvc-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17202122276/job/48794736037) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/48794736037) [(plain)](https://github.com/rust-lang/rust/commit/6bdc49a1b7f4f78122486a807a8c68ab243ccab2/checks/48794736037/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
13   --> $DIR/not-an-allocator.rs:2:11

18    |           ^^^^^ the trait `GlobalAlloc` is not implemented for `usize`
19    |
20 help: the trait `GlobalAlloc` is implemented for `System`
-   --> $SRC_DIR/std/src/sys/alloc/unix.rs:LL:COL
+   --> $SRC_DIR/std/src/sys/alloc/windows.rs:LL:COL
22    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
23 
24 error[E0277]: the trait bound `usize: GlobalAlloc` is not satisfied

30    |           ^^^^^ the trait `GlobalAlloc` is not implemented for `usize`
31    |
32 help: the trait `GlobalAlloc` is implemented for `System`
-   --> $SRC_DIR/std/src/sys/alloc/unix.rs:LL:COL
+   --> $SRC_DIR/std/src/sys/alloc/windows.rs:LL:COL
34    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
35 
36 error[E0277]: the trait bound `usize: GlobalAlloc` is not satisfied

42    |           ^^^^^ the trait `GlobalAlloc` is not implemented for `usize`
43    |
44 help: the trait `GlobalAlloc` is implemented for `System`
-   --> $SRC_DIR/std/src/sys/alloc/unix.rs:LL:COL
+   --> $SRC_DIR/std/src/sys/alloc/windows.rs:LL:COL
46    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
47 
48 error: aborting due to 4 previous errors

Note: some mismatched output was normalized before being compared
-   --> /rustc/FAKE_PREFIX\library\std\src\sys\alloc\windows.rs:161:1
---
To only update this specific test, also pass `--test-args allocator\not-an-allocator.rs`

error: 1 errors occurred comparing output.
status: exit code: 1
command: PATH="C:\a\rust\rust\build\aarch64-pc-windows-msvc\stage2\bin;C:\Program Files (x86)\Windows Kits\10\bin\arm64;C:\Program Files (x86)\Windows Kits\10\bin\10.0.26100.0\arm64;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.44.35207\bin\HostARM64\arm64;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.44.35207\bin\HostARM64\arm64;C:\a\rust\rust\build\aarch64-pc-windows-msvc\bootstrap-tools\aarch64-pc-windows-msvc\release\deps;C:\Program Files\Git\clangarm64\bin;C:\Program Files\Git\usr\bin;C:\Users\runneradmin\bin;C:\a\rust\rust\ninja;C:\a\rust\rust\citools\clang-rust\bin;C:\a\rust\rust\sccache;C:\aliyun-cli;C:\vcpkg;C:\Program Files (x86)\NSIS;C:\Program Files\Mercurial;C:\hostedtoolcache\windows\stack\3.7.1\x64;C:\mingw64\bin;C:\Program Files\dotnet;C:\Program Files\MySQL\MySQL Server 8.0\bin;C:\Program Files (x86)\R\R-4.4.2\bin\x64;C:\SeleniumWebDrivers\GeckoDriver;C:\SeleniumWebDrivers\EdgeDriver;C:\SeleniumWebDrivers\ChromeDriver;C:\Program Files (x86)\sbt\bin;C:\Program Files (x86)\GitHub CLI;C:\Program Files\Git\usr\bin;C:\Program Files (x86)\pipx_bin;C:\npm\prefix;C:\hostedtoolcache\windows\go\1.24.5\arm64\bin;C:\hostedtoolcache\windows\Python\3.12.10\arm64\Scripts;C:\hostedtoolcache\windows\Python\3.12.10\arm64;C:\hostedtoolcache\windows\Ruby\3.3.8\x64\bin;C:\Program Files\OpenSSL\bin;C:\tools\kotlinc\bin;C:\hostedtoolcache\windows\Java_Temurin-Hotspot_jdk\21.0.8-9.0\aarch64\bin;C:\Program Files (x86)\ImageMagick-7.1.2-Q16-HDRI;C:\Program Files\Microsoft SDKs\Azure\CLI2\wbin;C:\ProgramData\kind;C:\ProgramData\Chocolatey\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Windows\System32\OpenSSH;C:\Program Files\PowerShell\7;C:\Program Files\Microsoft\Web Platform Installer;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn;C:\Program Files\Microsoft SQL Server\150\Tools\Binn;C:\Program Files\dotnet;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit;C:\Program Files\Microsoft SQL Server\130\DTS\Binn;C:\Program Files\Microsoft SQL Server\140\DTS\Binn;C:\Program Files\Microsoft SQL Server\150\DTS\Binn;C:\Program Files\Microsoft SQL Server\160\DTS\Binn;C:\Strawberry\c\bin;C:\Strawberry\perl\site\bin;C:\Strawberry\perl\bin;C:\ProgramData\chocolatey\lib\pulumi\tools\Pulumi\bin;C:\Program Files\CMake\bin;C:\Tools\Ninja;C:\ProgramData\chocolatey\lib\maven\apache-maven-3.9.11\bin;C:\Program Files\LLVM\bin;C:\Program Files\nodejs;C:\Program Files\Git\cmd;C:\Program Files\Git\clangarm64\bin;C:\Program Files\Git\usr\bin;C:\Program Files\GitHub CLI;C:\tools\php;C:\Program Files (x86)\sbt\bin;C:\Program Files\Amazon\AWSCLIV2;C:\Program Files\Amazon\SessionManagerPlugin\bin;C:\Program Files\Amazon\AWSSAMCLI\bin;C:\Program Files\Microsoft SQL Server\130\Tools\Binn;C:\Users\runneradmin\.dotnet\tools;C:\Users\runneradmin\.cargo\bin;C:\Users\runneradmin\AppData\Local\Microsoft\WindowsApps" "C:\\a\\rust\\rust\\build\\aarch64-pc-windows-msvc\\stage2\\bin\\rustc.exe" "C:\\a\\rust\\rust\\tests\\ui\\allocator\\not-an-allocator.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=C:\\Users\\runneradmin\\.cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=C:\\a\\rust\\rust\\vendor" "--sysroot" "C:\\a\\rust\\rust\\build\\aarch64-pc-windows-msvc\\stage2" "--target=aarch64-pc-windows-msvc" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "C:\\a\\rust\\rust\\build\\aarch64-pc-windows-msvc\\test\\ui\\allocator\\not-an-allocator" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=C:\\a\\rust\\rust\\build\\aarch64-pc-windows-msvc\\native\\rust-test-helpers"
stdout: none
--- stderr -------------------------------
error[E0277]: the trait bound `usize: GlobalAlloc` is not satisfied
##[error]  --> C:\a\rust\rust\tests\ui\allocator\not-an-allocator.rs:2:11
   |
---
   | ------------------- in this attribute macro expansion
LL | static A: usize = 0;
   |           ^^^^^ the trait `GlobalAlloc` is not implemented for `usize`
   |
help: the trait `GlobalAlloc` is implemented for `System`
  --> /rustc/FAKE_PREFIX\library\std\src\sys\alloc\windows.rs:161:1
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

error[E0277]: the trait bound `usize: GlobalAlloc` is not satisfied
##[error]  --> C:\a\rust\rust\tests\ui\allocator\not-an-allocator.rs:2:11
   |
LL | #[global_allocator]
   | ------------------- in this attribute macro expansion
LL | static A: usize = 0;
   |           ^^^^^ the trait `GlobalAlloc` is not implemented for `usize`
   |
help: the trait `GlobalAlloc` is implemented for `System`
  --> /rustc/FAKE_PREFIX\library\std\src\sys\alloc\windows.rs:161:1
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

error[E0277]: the trait bound `usize: GlobalAlloc` is not satisfied
##[error]  --> C:\a\rust\rust\tests\ui\allocator\not-an-allocator.rs:2:11
   |
LL | #[global_allocator]
   | ------------------- in this attribute macro expansion
LL | static A: usize = 0;
   |           ^^^^^ the trait `GlobalAlloc` is not implemented for `usize`
   |
help: the trait `GlobalAlloc` is implemented for `System`
  --> /rustc/FAKE_PREFIX\library\std\src\sys\alloc\windows.rs:161:1
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0277`.
------------------------------------------
---

Some tests failed in compiletest suite=ui mode=ui host=aarch64-pc-windows-msvc target=aarch64-pc-windows-msvc
Bootstrap failed while executing `test --stage 2 --skip=compiler --skip=src`
Build completed unsuccessfully in 1:10:52
make: *** [Makefile:112: ci-msvc-py] Error 1
  local time: Mon Aug 25 08:48:04 CUT 2025
  network time: Mon, 25 Aug 2025 08:48:07 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/rust/actions/runs/17202122276/job/48800377733)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust/actions/runs/17202122276/job/48800377733","builder_name":"checks-actions"} -->

estebank commented:
@bors r=nnethercote

bors commented:
:pushpin: Commit c22331f3279c5f645caf3adac54406642850c526 has been approved by `nnethercote`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nnethercote c22331f3279c5f645caf3adac54406642850c526 -->
<!-- homu: {"type":"Approved","sha":"c22331f3279c5f645caf3adac54406642850c526","approver":"nnethercote","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit c22331f3279c5f645caf3adac54406642850c526 with merge 85abb8953ce3bf98a9ff685355f4c5afe78b3f09...
<!-- homu: {"type":"BuildStarted","head_sha":"c22331f3279c5f645caf3adac54406642850c526","merge_sha":"85abb8953ce3bf98a9ff685355f4c5afe78b3f09"} -->

rust-log-analyzer commented:
The job **`x86_64-mingw-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17219356461/job/48850858882) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/48850858882) [(plain)](https://github.com/rust-lang/rust/commit/85abb8953ce3bf98a9ff685355f4c5afe78b3f09/checks/48850858882/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

failures:

---- [ui] tests\ui\linkage-attr\raw-dylib\windows\dlltool-failed.rs stdout ----
\a\rust\rust\mingw64\bin\dlltool.exe
\AppData\Local\Temp\rustcpbgK0C\foo.dll_imports.def
\AppData\Local\Temp\rustcpbgK0C\foo.dll_imports.lib
\AppData\Local\Temp\rustcpbgK0C\foo.dll
\a\rust\rust\mingw64\bin\dlltool.exe
\AppData\Local\Temp\rustcpbgK0C\foo.dll_imports.def
Saved the actual stderr to `D:\a\rust\rust\build\x86_64-pc-windows-gnu\test\ui\linkage-attr\raw-dylib\windows\dlltool-failed\dlltool-failed.stderr`
diff of stderr:

1 error: Dlltool could not create import library with $DLLTOOL -d $DEF_FILE -D foo.dll -l $LIB_FILE $TARGET_MACHINE $ASM_FLAGS --no-leading-underscore $TEMP_PREFIX:
2        
3        $DLLTOOL: Syntax error in def file $DEF_FILE:1␍
+        
4 
5 error: aborting due to 1 previous error
6 


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args linkage-attr\raw-dylib\windows\dlltool-failed.rs`

error: 1 errors occurred comparing output.
status: exit code: 1
command: PATH="D:\a\rust\rust\build\x86_64-pc-windows-gnu\stage2\bin;D:\a\rust\rust\build\x86_64-pc-windows-gnu\bootstrap-tools\x86_64-pc-windows-gnu\release\deps;D:\a\rust\rust\ninja;D:\a\rust\rust\mingw64\bin;C:\msys64\usr\bin;D:\a\rust\rust\sccache;C:\Program Files\MongoDB\Server\7.0\bin;C:\vcpkg;C:\tools\zstd;C:\hostedtoolcache\windows\stack\3.7.1\x64;C:\cabal\bin;C:\ghcup\bin;C:\mingw64\bin;C:\Program Files\dotnet;C:\Program Files\MySQL\MySQL Server 8.0\bin;C:\Program Files\R\R-4.4.2\bin\x64;C:\SeleniumWebDrivers\GeckoDriver;C:\SeleniumWebDrivers\EdgeDriver;C:\SeleniumWebDrivers\ChromeDriver;C:\Program Files (x86)\sbt\bin;C:\Program Files (x86)\GitHub CLI;C:\Program Files\Git\bin;C:\Program Files (x86)\pipx_bin;C:\npm\prefix;C:\hostedtoolcache\windows\go\1.24.6\x64\bin;C:\hostedtoolcache\windows\Python\3.9.13\x64\Scripts;C:\hostedtoolcache\windows\Python\3.9.13\x64;C:\hostedtoolcache\windows\Ruby\3.3.9\x64\bin;C:\Program Files\OpenSSL\bin;C:\tools\kotlinc\bin;C:\hostedtoolcache\windows\Java_Temurin-Hotspot_jdk\17.0.16-8\x64\bin;C:\Program Files\ImageMagick-7.1.2-Q16-HDRI;C:\Program Files\Microsoft SDKs\Azure\CLI2\wbin;C:\ProgramData\kind;C:\ProgramData\Chocolatey\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Windows\System32\OpenSSH;C:\Program Files\PowerShell\7;C:\Program Files\Microsoft\Web Platform Installer;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn;C:\Program Files\Microsoft SQL Server\150\Tools\Binn;C:\Program Files\dotnet;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit;C:\Program Files (x86)\WiX Toolset v3.14\bin;C:\Program Files\Microsoft SQL Server\130\DTS\Binn;C:\Program Files\Microsoft SQL Server\140\DTS\Binn;C:\Program Files\Microsoft SQL Server\150\DTS\Binn;C:\Program Files\Microsoft SQL Server\160\DTS\Binn;C:\ProgramData\chocolatey\lib\pulumi\tools\Pulumi\bin;C:\Program Files\CMake\bin;C:\Strawberry\c\bin;C:\Strawberry\perl\site\bin;C:\Strawberry\perl\bin;C:\ProgramData\chocolatey\lib\maven\apache-maven-3.9.11\bin;C:\Program Files\Microsoft Service Fabric\bin\Fabric\Fabric.Code;C:\Program Files\Microsoft SDKs\Service Fabric\Tools\ServiceFabricLocalClusterManager;C:\Program Files\nodejs;C:\Program Files\Git\cmd;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Program Files\GitHub CLI;C:\tools\php;C:\Program Files (x86)\sbt\bin;C:\Program Files\Amazon\AWSCLIV2;C:\Program Files\Amazon\SessionManagerPlugin\bin;C:\Program Files\Amazon\AWSSAMCLI\bin;C:\Program Files\Microsoft SQL Server\130\Tools\Binn;C:\Program Files\mongosh;C:\Program Files\LLVM\bin;C:\Program Files (x86)\LLVM\bin;C:\Users\runneradmin\.dotnet\tools;C:\Users\runneradmin\.cargo\bin;C:\Users\runneradmin\AppData\Local\Microsoft\WindowsApps" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\stage2\\bin\\rustc.exe" "D:\\a\\rust\\rust\\tests\\ui\\linkage-attr\\raw-dylib\\windows\\dlltool-failed.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=C:\\Users\\runneradmin\\.cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=D:\\a\\rust\\rust\\vendor" "--sysroot" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\stage2" "--target=x86_64-pc-windows-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\test\\ui\\linkage-attr\\raw-dylib\\windows\\dlltool-failed" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\native\\rust-test-helpers" "--crate-type" "lib" "--emit" "link"
stdout: none
--- stderr -------------------------------
error: Dlltool could not create import library with D:\a\rust\rust\mingw64\bin\dlltool.exe -d C:\Users\RUNNER~1\AppData\Local\Temp\rustcpbgK0C\foo.dll_imports.def -D foo.dll -l C:\Users\RUNNER~1\AppData\Local\Temp\rustcpbgK0C\foo.dll_imports.lib -m i386:x86-64 -f --64 --no-leading-underscore --temp-prefix C:\Users\RUNNER~1\AppData\Local\Temp\rustcpbgK0C\foo.dll:
       
       D:\a\rust\rust\mingw64\bin\dlltool.exe: Syntax error in def file C:\Users\RUNNER~1\AppData\Local\Temp\rustcpbgK0C\foo.dll_imports.def:1␍
       

error: aborting due to 1 previous error
------------------------------------------

---

Some tests failed in compiletest suite=ui mode=ui host=x86_64-pc-windows-gnu target=x86_64-pc-windows-gnu
Bootstrap failed while executing `test --stage 2 --skip=compiler --skip=src`
Build completed unsuccessfully in 1:45:06
make: *** [Makefile:123: ci-mingw-x] Error 1
  local time: Mon Aug 25 21:54:27 CUT 2025
  network time: Mon, 25 Aug 2025 21:54:28 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

bors commented:
:hourglass: Testing commit 473078eae43d3826705b8a588e19b82f94336cc4 with merge 6f6a43f5328af7faed02e2cfb24ca510410c98d1...
<!-- homu: {"type":"BuildStarted","head_sha":"473078eae43d3826705b8a588e19b82f94336cc4","merge_sha":"6f6a43f5328af7faed02e2cfb24ca510410c98d1"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #146614) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:hourglass: Testing commit cbe7ba6c825521e8a04e927260ece36400ad4769 with merge e587c0ea66f2f26b956e14629747f0a3b26fcd23...
<!-- homu: {"type":"BuildStarted","head_sha":"cbe7ba6c825521e8a04e927260ece36400ad4769","merge_sha":"e587c0ea66f2f26b956e14629747f0a3b26fcd23"} -->

rust-log-analyzer commented:
The job **`x86_64-mingw-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18078941391/job/51439470641) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/51439470641) [(plain)](https://github.com/rust-lang/rust/commit/e587c0ea66f2f26b956e14629747f0a3b26fcd23/checks/51439470641/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] tests\ui\linkage-attr\raw-dylib\windows\dlltool-failed.rs stdout ----
Saved the actual stderr to `D:\a\rust\rust\build\x86_64-pc-windows-gnu\test\ui\linkage-attr\raw-dylib\windows\dlltool-failed\dlltool-failed.stderr`
diff of stderr:

1 error: dlltool could not create import library with $DLLTOOL -d $DEF_FILE -D foo.dll -l $LIB_FILE $TARGET_MACHINE $ASM_FLAGS --no-leading-underscore $TEMP_PREFIX:
2        
3        $DLLTOOL: Syntax error in def file $DEF_FILE:1␍
- 
+        
5 
6 error: aborting due to 1 previous error
7 
---
To only update this specific test, also pass `--test-args linkage-attr\raw-dylib\windows\dlltool-failed.rs`

error: 1 errors occurred comparing output.
status: exit code: 1
command: PATH="D:\a\rust\rust\build\x86_64-pc-windows-gnu\stage2\bin;D:\a\rust\rust\build\x86_64-pc-windows-gnu\bootstrap-tools\x86_64-pc-windows-gnu\release\deps;D:\a\rust\rust\ninja;D:\a\rust\rust\mingw64\bin;C:\msys64\usr\bin;D:\a\rust\rust\sccache;C:\Program Files\MongoDB\Server\7.0\bin;C:\vcpkg;C:\tools\zstd;C:\hostedtoolcache\windows\stack\3.7.1\x64;C:\cabal\bin;C:\ghcup\bin;C:\mingw64\bin;C:\Program Files\dotnet;C:\Program Files\MySQL\MySQL Server 8.0\bin;C:\Program Files\R\R-4.4.2\bin\x64;C:\SeleniumWebDrivers\GeckoDriver;C:\SeleniumWebDrivers\EdgeDriver;C:\SeleniumWebDrivers\ChromeDriver;C:\Program Files (x86)\sbt\bin;C:\Program Files (x86)\GitHub CLI;C:\Program Files\Git\bin;C:\Program Files (x86)\pipx_bin;C:\npm\prefix;C:\hostedtoolcache\windows\go\1.24.7\x64\bin;C:\hostedtoolcache\windows\Python\3.9.13\x64\Scripts;C:\hostedtoolcache\windows\Python\3.9.13\x64;C:\hostedtoolcache\windows\Ruby\3.3.9\x64\bin;C:\Program Files\OpenSSL\bin;C:\tools\kotlinc\bin;C:\hostedtoolcache\windows\Java_Temurin-Hotspot_jdk\17.0.16-8\x64\bin;C:\Program Files\ImageMagick-7.1.2-Q16-HDRI;C:\Program Files\Microsoft SDKs\Azure\CLI2\wbin;C:\ProgramData\kind;C:\ProgramData\Chocolatey\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Windows\System32\OpenSSH;C:\Program Files\PowerShell\7;C:\Program Files\Microsoft\Web Platform Installer;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn;C:\Program Files\Microsoft SQL Server\150\Tools\Binn;C:\Program Files\dotnet;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit;C:\Program Files (x86)\WiX Toolset v3.14\bin;C:\Program Files\Microsoft SQL Server\130\DTS\Binn;C:\Program Files\Microsoft SQL Server\140\DTS\Binn;C:\Program Files\Microsoft SQL Server\150\DTS\Binn;C:\Program Files\Microsoft SQL Server\160\DTS\Binn;C:\ProgramData\chocolatey\lib\pulumi\tools\Pulumi\bin;C:\Program Files\CMake\bin;C:\Strawberry\c\bin;C:\Strawberry\perl\site\bin;C:\Strawberry\perl\bin;C:\ProgramData\chocolatey\lib\maven\apache-maven-3.9.11\bin;C:\Program Files\Microsoft Service Fabric\bin\Fabric\Fabric.Code;C:\Program Files\Microsoft SDKs\Service Fabric\Tools\ServiceFabricLocalClusterManager;C:\Program Files\nodejs;C:\Program Files\Git\cmd;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Program Files\GitHub CLI;C:\tools\php;C:\Program Files (x86)\sbt\bin;C:\Program Files\Amazon\AWSCLIV2;C:\Program Files\Amazon\SessionManagerPlugin\bin;C:\Program Files\Amazon\AWSSAMCLI\bin;C:\Program Files\Microsoft SQL Server\130\Tools\Binn;C:\Program Files\mongosh;C:\Program Files\LLVM\bin;C:\Program Files (x86)\LLVM\bin;C:\Users\runneradmin\.dotnet\tools;C:\Users\runneradmin\.cargo\bin;C:\Users\runneradmin\AppData\Local\Microsoft\WindowsApps" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\stage2\\bin\\rustc.exe" "D:\\a\\rust\\rust\\tests\\ui\\linkage-attr\\raw-dylib\\windows\\dlltool-failed.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=C:\\Users\\runneradmin\\.cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=D:\\a\\rust\\rust\\vendor" "--sysroot" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\stage2" "--target=x86_64-pc-windows-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\test\\ui\\linkage-attr\\raw-dylib\\windows\\dlltool-failed" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\native\\rust-test-helpers" "--crate-type" "lib" "--emit" "link"
stdout: none
--- stderr -------------------------------
error: dlltool could not create import library with D:\a\rust\rust\mingw64\bin\dlltool.exe -d C:\Users\RUNNER~1\AppData\Local\Temp\rustcIYiLQQ\foo.dll_imports.def -D foo.dll -l C:\Users\RUNNER~1\AppData\Local\Temp\rustcIYiLQQ\foo.dll_imports.lib -m i386:x86-64 -f --64 --no-leading-underscore --temp-prefix C:\Users\RUNNER~1\AppData\Local\Temp\rustcIYiLQQ\foo.dll:
       
       D:\a\rust\rust\mingw64\bin\dlltool.exe: Syntax error in def file C:\Users\RUNNER~1\AppData\Local\Temp\rustcIYiLQQ\foo.dll_imports.def:1␍
       

error: aborting due to 1 previous error
------------------------------------------

---

Some tests failed in compiletest suite=ui mode=ui host=x86_64-pc-windows-gnu target=x86_64-pc-windows-gnu
Bootstrap failed while executing `test --stage 2 --skip=compiler --skip=src`
Build completed unsuccessfully in 1:48:23
make: *** [Makefile:123: ci-mingw-x] Error 1
  local time: Sun Sep 28 21:38:03 CUT 2025
  network time: Sun, 28 Sep 2025 21:38:04 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/rust/actions/runs/18078941391/job/51442333732)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust/actions/runs/18078941391/job/51442333732","builder_name":"checks-actions"} -->

estebank committed:
Fix `dlltool-failed` test

estebank committed:
Account for `#[rustc_reservation_impl]` when pointing at potential impls

bors commented:
:pushpin: Commit 3b3dc7bd8cefd98af8ac38b8530696770f37b0be has been approved by `nnethercote`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nnethercote 3b3dc7bd8cefd98af8ac38b8530696770f37b0be -->
<!-- homu: {"type":"Approved","sha":"3b3dc7bd8cefd98af8ac38b8530696770f37b0be","approver":"nnethercote","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

estebank committed:
Ignore file length lint for now

bors commented:
:hourglass: Testing commit 32aa59939cf0c60dacf417376a134c668f4b48be with merge bd3ac0330018c23b111bbee176f32c377be7b319...
<!-- homu: {"type":"BuildStarted","head_sha":"32aa59939cf0c60dacf417376a134c668f4b48be","merge_sha":"bd3ac0330018c23b111bbee176f32c377be7b319"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/19002028343/job/54275492847)
Approved by: nnethercote
Pushing bd3ac0330018c23b111bbee176f32c377be7b319 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nnethercote","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/19002028343/job/54275492847"},"merge_sha":"bd3ac0330018c23b111bbee176f32c377be7b319"} -->
